io/pci/pci_hotplug.py

```diff
@@ -131,8 +114,29 @@ def hotplug_add(self, slot, pci_addr):
         genio.write_file("/sys/bus/pci/slots/%s/power" % slot, "1")
 
         def is_added():
+            """
+            Returns True if pci device is added, False otherwise.
+            """
             if pci_addr not in pci.get_pci_addresses():
                 return False
             return True
 
-        return wait.wait_for(is_added, timeout=10) or False
+        def is_recovered():
+            """
+            Compares current endpoint devices in pci address with
+            `pre` value, and returns True if equals pre.
+            False otherwise.
+            """
+            post = len(pci.get_disks_in_pci_address(pci_addr))
+            post += len(pci.get_nics_in_pci_address(pci_addr))
+            self.log.debug("Pre: %d,  Post: %d",
+                           self.end_devices[pci_addr], post)
+            if post == self.end_devices[pci_addr]:
+                return True
```

Comment:
@narasimhan-v count of disk doesn't help to verify exactly here.
disks may appear bak but they may online or offline. So I feel checking the status of each disk will be better option than only checking the count.

let me know if you need any inputs in disk status check.

Reply:
@Naresh-ibm Sure, but we are not checking just disks here. We are checking all devices count.
Can you explain the disk online / offline part more please ?

Reply:
disks_list = pci.get_disks_in_pci_address(pci_addr)
for disk in disks_list:
     status = multipath.get_path_status('disk')
     if status[0] == 'active' or status[2] == 'ready':
          err_disk.append(disk)

if err_disk:
    self.fail("following disk failed to ")


vise versa for adding.

Reply:
@Naresh-ibm Are you under the assumption that this is being done for only multipath devices ?
We do this test for all types of pci devices: network, nvme, fc, etc.

Reply:
@narasimhan-v Not under assumption of multipath.
Add the following condition which will work
if disk_list:
    for disk in disks_list:
         status = multipath.get_path_status('disk')
         if status[0] == 'active' or status[2] == 'ready':
              err_disk.append(disk)

if err_disk:
self.fail("following disk failed to ")

apply all disk status checks.

if it is n/w adapter, obviously it will not have any disk and it will skip this checks.
left thing is nvme and ipr/lsi. checking for LSI. let me know on nvme part.

if this is not going work with nvme and ipr then lets add a field with FC adapter in yaml and then accordingly we will verify . As multipath adapters are more sensitive to this operations. Most of the time multipath does not recover properly on dlpar/pci_hotplug. so I feel adding this will make the script more robust.

if above logic works fine for nvme and ipr/LSI then lets include this here.

Reply:
@Naresh-ibm I guess we take that up as an enhancement, and you can add it for FC. And other components can be added as needed by people interested.

Reply:
genioread(/sys/block/sda/device/state)

check this path for individual path states. I have verified for LSI and FC. and I think same will applicable for all block devices.

Reply:
@Naresh-ibm Sure.
1. This is definitely good to do, but we need generic solutions for all types. If not possible, we can handle it for disks alone.
2. This disk state should go into avocado utils, and then we can use in tests. Can you send a patch for the same please ?

And count of disks and network devices is the primary check to see if all devices recovered. We can discuss that here, and the enhancement can come later. Let me know.

Reply:
Documenting:

```
grep . /sys/block/*/device/state
/sys/block/nvme1n1/device/state:live
/sys/block/nvme1n2/device/state:live
/sys/block/sda/device/state:running
```

Reply:
@narasimhan-v  sure, I will send a patch, can you point me out which file I need add this? do we have any specific file like multipath for block devices?

Reply:
since this is disk status, this should go into: https://github.com/avocado-framework/avocado/blob/master/avocado/utils/disk.py

If you can cover it for network also, then it can be considered as pci util and can go into: 
https://github.com/avocado-framework/avocado/blob/master/avocado/utils/pci.py

Reply:
@Naresh-ibm Considering the above, requesting to re-review the patch.